### PR TITLE
Fix PassLocalVarType handling of If statements and reduce

### PR DESCRIPTION
**Pull request by Stagno**
## Technical Description

`PassLocalVarType` was not considering the accesses inside the conditional of an If statement and reductions. This PR fixes this.

### CODE REVIEW

**Comment by Stagno** [comment]
launch jenkins

**Comment by Stagno** [comment]
launch jenkins

**Comment by Stagno** [comment]
launch jenkins

File: dawn/src/dawn/Optimizer/PassLocalVarType.cpp [hunk 1]
```diff
@@ -91,29 +99,63 @@ class VarTypeFinder : public ast::ASTVisitorForwarding {
...
+
+  // Construct a LocalVariableType from a field accessed within an assignment to / declaration of a
+  // local variable.
+  iir::LocalVariableType inferLocalVarTypeFromField(int fieldAccessID) {
+    iir::LocalVariableType type;
+
+    if(sir::dimension_isa<sir::CartesianFieldDimension>(
+           metadata_.getFieldDimensions(fieldAccessID).getHorizontalFieldDimension())) {
+      // Cartesian case
+      type = iir::LocalVariableType::OnIJ;
```
Comment (havogt) [inline dawn/src/dawn/Optimizer/PassLocalVarType.cpp]: return here

File: dawn/src/dawn/Optimizer/PassLocalVarType.cpp [hunk 2]
```diff
@@ -91,29 +99,63 @@ class VarTypeFinder : public ast::ASTVisitorForwarding {
...
+    } else {
+      // Unstructured case, convert from location type
+      type =
+          localVarTypeFromLocationType(metadata_.getDenseLocationTypeFromAccessID(fieldAccessID));
+    }
+    return type;
```
Comment (havogt) [inline dawn/src/dawn/Optimizer/PassLocalVarType.cpp]: not here

File: dawn/src/dawn/Optimizer/PassLocalVarType.cpp [hunk 3]
```diff
@@ -91,29 +99,63 @@ class VarTypeFinder : public ast::ASTVisitorForwarding {
...
+  }
+  // Records that to write into `destinationVariable` we need to access `accessedVariable`.
+  // `sourceLocation` is where this happens.
+  void recordVariablePair(int accessedVariable, int destinationVariable,
+                          SourceLocation sourceLocation) {
+    // No self-dependencies
+    if(accessedVariable != destinationVariable) {
+
+      // If the set of dependers for this variable is not there, it means the variable hasn't been
```
Comment (havogt) [inline dawn/src/dawn/Optimizer/PassLocalVarType.cpp]: remove comment, it's already in the assert message

File: dawn/src/dawn/Optimizer/PassLocalVarType.cpp [hunk 4]
```diff
@@ -91,29 +99,63 @@ class VarTypeFinder : public ast::ASTVisitorForwarding {
...
+      // declared
+      DAWN_ASSERT_MSG(dependencyMap_.count(accessedVariable),
+                      "Variable accessed before being declared.");
+      // Need to register a dependency. Pair to be added is from dependency (variable accessed in
```
Comment (havogt) [inline dawn/src/dawn/Optimizer/PassLocalVarType.cpp]: if you think this comment is valuable, replace it by a `registerDependency()` function

File: dawn/src/dawn/Optimizer/PassLocalVarType.cpp [hunk 5]
```diff
@@ -91,29 +99,63 @@ class VarTypeFinder : public ast::ASTVisitorForwarding {
...
+      // the rhs) to dependent (lhs).
+      dependencyMap_.at(accessedVariable).insert(destinationVariable);
+
+      // Retrieve the type of the accessed variable
+      const iir::LocalVariableData& data =
+          metadata_.getLocalVariableDataFromAccessID(accessedVariable);
+      DAWN_ASSERT_MSG(data.isTypeSet(), "Variable accessed before being declared.");
+      iir::LocalVariableType newType = data.getType();
+
+      // Update the type of the current variable with `newType`
```
Comment (havogt) [inline dawn/src/dawn/Optimizer/PassLocalVarType.cpp]: remove comment

File: dawn/src/dawn/Optimizer/PassLocalVarType.cpp [hunk 6]
```diff
@@ -123,44 +165,64 @@ class VarTypeFinder : public ast::ASTVisitorForwarding {
...
-        // Update the type of the current variable with `newType`
-        updateVariableType(*curVarID_, newType, expr->getSourceLocation());
+      recordVariablePair(accessedVariableID, *curVarID_, expr->getSourceLocation());
+    }
+  }
+  void visit(const std::shared_ptr<iir::IfStmt>& ifStmt) override {
+    // If in the conditional expression we are accessing a field, it means that every statement in
+    // the then and else blocks will have a read-dependency on such field.
+
+    for(const auto& readAccess :
```
Comment (havogt) [inline dawn/src/dawn/Optimizer/PassLocalVarType.cpp]: You could consider using structured bindings here
```
for(const auto& [accessID, unused] : ifStmt->getCondStmt()->getData<iir::IIRStmtData>().CallerAccesses->getReadAccesses())
```
but also not very nice to have the `unused` around.

File: dawn/src/dawn/Optimizer/PassLocalVarType.cpp [hunk 7]
```diff
@@ -169,12 +231,31 @@ class VarTypeFinder : public ast::ASTVisitorForwarding {
...
+  }
   void visit(const std::shared_ptr<iir::VarDeclStmt>& stmt) override {
-    // Assume scalar if nothing proves the opposite during the visit
+    // Set curVarID_ to current variable being processed
     curVarID_ = iir::getAccessID(stmt);
-    updateVariableType(*curVarID_, iir::LocalVariableType::Scalar, stmt->getSourceLocation());
+    // If we are inside an if-statement and the if condition accesses a field, we already have a
+    // variable type inferred. Otherwise, assume scalar as starting type. In any case we need
+    // to do the visit.
+    const iir::LocalVariableType startingType =
```
Comment (havogt) [inline dawn/src/dawn/Optimizer/PassLocalVarType.cpp]: remove code duplication

File: dawn/src/dawn/Optimizer/PassLocalVarType.cpp (hunk 4 above)
Comment (Stagno) [reply dawn/src/dawn/Optimizer/PassLocalVarType.cpp]: I think if one reads the description of `dependencyMap_`, then the code is already self-explanatory. I'll remove the comment.

File: dawn/src/dawn/Optimizer/PassLocalVarType.cpp (hunk 6 above)
Comment (Stagno) [reply dawn/src/dawn/Optimizer/PassLocalVarType.cpp]: I guess there's no elegant way to iterate through the keys only ...

**Comment by Stagno** [comment]
launch jenkins

File: dawn/src/dawn/Optimizer/PassLocalVarType.cpp (hunk 6 above)
Comment (havogt) [reply dawn/src/dawn/Optimizer/PassLocalVarType.cpp]: yes...

**Pull request merged**